syllabus/task_space/task_space.py

```diff
@@ -11,34 +11,179 @@ def __init__(self, gym_space: Union[Space, int], tasks=None):
-            encoder = lambda task: self._encode_map[task] if task in self._encode_map else None
-            decoder = lambda task: self._decode_map[task] if task in self._decode_map else None
+            encode_map = {task: i for i, task in enumerate(tasks)}
+            decode_map = {i: task for i, task in enumerate(tasks)}
+            encoder = lambda task: encode_map[task] if task in encode_map else None
+            decoder = lambda task: decode_map[task] if task in decode_map else None
         elif isinstance(space, Box):
             encoder = lambda task: task if space.contains(np.asarray(task, dtype=space.dtype)) else None
             decoder = lambda task: task if space.contains(np.asarray(task, dtype=space.dtype)) else None
         elif isinstance(space, Tuple):
-            for i, task in enumerate(tasks):
-                assert self.count_tasks(space.spaces[i]) == len(task), "Each task must have number of components equal to Tuple space length. Got {len(task)} components and space length {self.count_tasks(space.spaces[i])}."
+           
+            assert len(space.spaces) == len(tasks), f"Number of task ({len(space.spaces)})must match options in Tuple ({len(tasks)})"
             results = [list(self._make_task_encoder(s, t)) for (s, t) in zip(space.spaces, tasks)]
             encoders = [r[0] for r in results]
             decoders = [r[1] for r in results]
             encoder = lambda task: [e(t) for e, t in zip(encoders, task)]
             decoder = lambda task: [d(t) for d, t in zip(decoders, task)]
+
+        elif isinstance(space, MultiDiscrete):
+            assert len(space.nvec) == len(tasks), f"Number of steps in a tasks ({len(space.nvec)}) must match number of discrete options ({len(tasks)})"
+            
+            combinations = [p for p in itertools.product(*tasks)]
+            encode_map = {task: i for i, task in enumerate(combinations)}
+            decode_map = {i: task for i, task in enumerate(combinations)}
+    
+            encoder = lambda task: encode_map[task] if task in encode_map else None
+            decoder = lambda task: decode_map[task] if task in decode_map else None
+
+        elif isinstance(space, Dict):
+
+            # Function flatten() takes a nested dictionary as input and flattens it to have each key in
+            # root's level. Keys whos values are dictionaries store the length of keys. 
+            # This function will help create a mapping dictionary that can store the encoder that will use the 
+            # space and tasks value based on keys.
+
+            def flatten(d,out):
+                for (key,value) in d.items() :
+                    if ( isinstance(value, dict) or isinstance(value, Dict))  :
+                        out.update({key : len(value)})
+                        out.update(flatten(value, {}))
+                    else:
+                        out[key] = value
+                return out
```

Comment:
```suggestion
            def flatten(d):
            """ Add documentation for flatten method here. """
                out = {}
                for (key,value) in d.items() :
                    if ( isinstance(value, dict) or isinstance(value, Dict))  :
                        out.update({key : len(value)})
                        out.update(flatten(value, {}))
                    else:
                        out[key] = value
                return out
```

I'd prefer not to use the functional approach here, so just define out within the method. Also, what happens if two subdictionaries contain the same key? For example {"Attack with": {"Melee": 0, "Range": 1}, "Level up": {"Melee": 2, "Range": 3}}

Reply:
Also why store the length under the original key? What if multiple sub-dictionaries have the same length?